### DISCUSSION TITLE
[Question][Hw5] Why does Memory.find return Value.bottom when variable is not found?

### DISCUSSION THREAD

**Discussion by AnHaechan** [prosyslab-classroom/cs348-information-security#609] [Q&A]
`Memory.find` is defined as following:
`let find x m = try M.find x m with Not_found -> Value.bottom`
but I'm curious about the purpose of returning `Value.bottom` on `Not_found`.

By that design choice, we cannot distinguish whether the `x is not in m` or `m(x)=ㅗ` .
And this became problem at `Sematics.Make.eval`.
Because we're using `llvaue (of instruction)` as the `x : Variable.t` (key of the map),
given `eval x m`, we don't know whether
(1) it is first evaluating instruction x
or (2) it is called during evaluation of other instruction such as y=x+1.

For example,
```
int x = 3;
int y;
if (x<0){
  y = x+1;
}
```
At node `int x=3`, `mem[x|->eval(x, mem)]` will be done.
In this case, x is evaluated very first, so `x not in mem` thus `find x mem` will return bottom.
At node `y = x+1`, `mem[y|->eval(x+1, mem)]` will be done. Inside `eval(x+1, mem)`, `eval(x,mem)` will be called.
In this case, `mem(x) = ㅗ` because of the if conditional, thus `find x mem` will return bottom.

In the point of view of `eval`, how can we distinguish above two cases?
I've temporarily added new function in module `Memory` to see if `x is m or not`,
but I think there must be a reason for the provided implementation of `Memory.find`.

**Comment by shyngys-aitkazinov** [comment]
I guess it is because the bottom means the  null set meaning that not exist or not found
